chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java

```diff
@@ -1370,51 +1372,111 @@ public synchronized void loadChunks(TaskTracker taskTracker, World world, Collec
       for (ChunkPosition cp : nonEmptyChunks) {
         // Finalize grass and foliage textures.
         // 3x3 box blur.
-        for (int x = 0; x < 16; ++x) {
-          for (int z = 0; z < 16; ++z) {
-
-            int nsum = 0;
-            float[] grassMix = {0, 0, 0};
-            float[] foliageMix = {0, 0, 0};
-            float[] waterMix = {0, 0, 0};
-            for (int sx = x - 1; sx <= x + 1; ++sx) {
-              int wx = cp.x * 16 + sx;
-              for (int sz = z - 1; sz <= z + 1; ++sz) {
-                int wz = cp.z * 16 + sz;
-
-                ChunkPosition ccp = ChunkPosition.get(wx >> 4, wz >> 4);
-                if (nonEmptyChunks.contains(ccp)) {
-                  nsum += 1;
-                  Biome biome = biomePalette.get(biomePaletteIdxMap.get(wx, wz));
-                  float[] grassColor = biome.grassColorLinear;
-                  grassMix[0] += grassColor[0];
-                  grassMix[1] += grassColor[1];
-                  grassMix[2] += grassColor[2];
-                  float[] foliageColor = biome.foliageColorLinear;
-                  foliageMix[0] += foliageColor[0];
-                  foliageMix[1] += foliageColor[1];
-                  foliageMix[2] += foliageColor[2];
-                  float[] waterColor = biome.waterColorLinear;
-                  waterMix[0] += waterColor[0];
-                  waterMix[1] += waterColor[1];
-                  waterMix[2] += waterColor[2];
+        if(using3dBiomes) {
+          for (int sectionY = yMin >> 4; sectionY < (yMax - 1 >> 4) + 1; sectionY++) {
+            for (int x = 0; x < 16; ++x) {
+              for (int z = 0; z < 16; ++z) {
+                for (int y = 0; y < 16; y++) {
+                  int nsum = 0;
+
+                  float[] grassMix = {0, 0, 0};
+                  float[] foliageMix = {0, 0, 0};
+                  float[] waterMix = {0, 0, 0};
+                  for (int sx = x - 1; sx <= x + 1; ++sx) {
+                    int wx = cp.x * 16 + sx;
+                    for (int sz = z - 1; sz <= z + 1; ++sz) {
+                      int wz = cp.z * 16 + sz;
+                      for (int sy = y - 1; sy < y + 1; sy++) {
+                        int wy = sectionY * 16 + sy;
+                        ChunkPosition ccp = ChunkPosition.get(wx >> 4, wz >> 4);
+                        if (nonEmptyChunks.contains(ccp)) {
+                          nsum += 1;
+                          Integer id = biomePaletteIdxStructure.get(wx, wy, wz);
+                          if (id == null) {
```

Comment:
If I'm not mistaken, get returns an int so the boxen and test for null is not needed